nomad/state/state_store.go

```diff
@@ -3177,6 +3177,55 @@ func (s *StateStore) EvalByID(ws memdb.WatchSet, id string) (*structs.Evaluation
 	return nil, nil
 }
 
+// EvalsRelatedToID is used to retrieve the evals that are related (next,
+// previous, or blocked) to the provided eval ID.
+func (s *StateStore) EvalsRelatedToID(ws memdb.WatchSet, id string) ([]*structs.EvaluationStub, error) {
+	txn := s.db.ReadTxn()
+
+	raw, err := txn.First("evals", "id", id)
+	if err != nil {
+		return nil, fmt.Errorf("eval lookup failed: %v", err)
+	}
+	if raw == nil {
+		return nil, nil
+	}
+	eval := raw.(*structs.Evaluation)
+
+	relatedEvals := []*structs.EvaluationStub{}
+	todo := eval.RelatedIDs()
+	done := map[string]bool{
+		eval.ID: true, // don't placing the requested eval in the related list.
+	}
+
+	for len(todo) > 0 {
+		// Pop the first value from the todo list.
+		current := todo[0]
+		todo = todo[1:]
+		if current == "" {
+			continue
+		}
+
+		// Skip value if we already have it in the results.
+		if _, ok := done[current]; ok {
+			continue
+		}
+
+		eval, err := s.EvalByID(ws, current)
```

Comment:
This call will add the eval watch channel to the watchset, which means we can have multiple channels in `ws`. I don't know if there are any performance implications in this approach.

A alternative would be to set the watchset using the job ID prefix. This would use a single watch channel, but may fire in situations where an eval for another with the same prefix changes.